sklearn/preprocessing/_polynomial.py

```diff
@@ -1008,16 +1062,37 @@ def transform(self, X):
                 for j in range(degree):
                     mask = X[:, i] < xmin
                     if np.any(mask):
-                        XBS[mask, i * n_splines + j] = (
-                            f_min[j] + (X[mask, i] - xmin) * fp_min[j]
-                        )
+                        linear_extr = f_min[j] + (X[mask, i] - xmin) * fp_min[j]
+                        if use_sparse:
+                            # Note: See comment about SparseEfficiencyWarning above.
+                            if not sparse.isspmatrix_lil(XBS_sparse):
+                                XBS_sparse = XBS_sparse.tolil()
+                            XBS_sparse[mask, j] = linear_extr
+                        else:
+                            XBS[mask, i * n_splines + j] = linear_extr
 
                     mask = X[:, i] > xmax
                     if np.any(mask):
                         k = n_splines - 1 - j
-                        XBS[mask, i * n_splines + k] = (
-                            f_max[k] + (X[mask, i] - xmax) * fp_max[k]
-                        )
+                        linear_extr = f_max[k] + (X[mask, i] - xmax) * fp_max[k]
+                        if use_sparse:
+                            # Note: See comment about SparseEfficiencyWarning above.
+                            if not sparse.isspmatrix_lil(XBS_sparse):
+                                XBS_sparse = XBS_sparse.tolil()
+                            XBS_sparse[mask, k : k + 1] = linear_extr[:, None]
+                        else:
+                            XBS[mask, i * n_splines + k] = linear_extr
+
+            if use_sparse:
+                if not sparse.isspmatrix_csr(XBS_sparse):
+                    XBS_sparse = XBS_sparse.tocsr()
+                output_list.append(XBS_sparse)
+
+        if use_sparse:
+            XBS = sparse.hstack(output_list)
```

Comment:
Before `hstack`-ing, should we add some extra logic regarding the integral `dtype` used for `indices` and `indptr` as done in https://github.com/scikit-learn/scikit-learn/pull/23731/?

https://github.com/scikit-learn/scikit-learn/blob/c973d642893f290811cd416b0b89bf1a8afd4011/sklearn/preprocessing/_polynomial.py#L414-L433

Reply:
Something like https://github.com/scikit-learn/scikit-learn/pull/24145/commits/e4b010d8bae4038a5444afee5b5613f65ad6d3ce?

Reply:
e4b010d looks good to me!

As proposed in https://github.com/scikit-learn/scikit-learn/pull/24145/files#r1007881886 just above, I would just nitpickingly add:
 - a `TODO` comment indicating to remove the check when we bump the minimum version of SciPy
  - a final full stop for the error message